### ISSUE TITLE
Test and incorporate new script for Collections to the NSSDCA Delivery Process

### ISSUE THREAD

**Issue by c-suh** [NASA-PDS/pdsen-operations#58]
Courtesy of @rchenatjpl.

Origin ticket: https://github.jpl.nasa.gov/PDSEN/data-engineering/issues/131

**Comment by jordanpadams** [comment]
@c-suh the process should also include re-registering the collection. ping @rchenatjpl for how to do that.

**Comment by rchenatjpl** [comment]
Hi, Catherine, you need python3 installed at /usr/bin/python3. If you don't, sorry, I wasn't expecting other people to run this - you can just ask me to run it, 'cause one of us will need to ingest the new collection anyway. Then it's
  makeCollection.py directoryWithPDS4Labels
which creates a Collection.xml and .csv for each unique collection encountered in the input directory. If the directory has an old collection.xml, it'll create the next version and bump the version number and write the difference in the modification history.

**Comment by jordanpadams** [comment]
@c-suh maybe add this into this repo as a script since we already have some other python scripts here?

**Comment by jordanpadams** [comment]
and you can use the virtualenv we have for this repo as well. if that makes any sense...

**Comment by c-suh** [comment]
@jordanpadams will do on adding the script after testing. And, yes, the virtualenv makes sense (you already documented it in this repo's README for another script).

@rchenatjpl okey doke; thank you for the information and offer. I will give it a go and let you know how it goes and if I'll need more of your help.

**Comment by rchenatjpl** [comment]
Here's a modified version that avoids mv /tmp/somethig, which failed for me on pds-gamms
[makeCollection.py.zip](https://github.com/NASA-PDS/pdsen-operations/files/5964133/makeCollection.py.zip)

**Comment by rchenatjpl** [comment]
And another version with prettier output since we'll be running this more often
[makeCollection.py.zip](https://github.com/NASA-PDS/pdsen-operations/files/5968923/makeCollection.py.zip)

**Comment by c-suh** [comment]
@rchenatjpl thank you for the updates! I ran this against the latest batch of 30 bundles (yet to be validated) and placed the resulting Collection*_v2.0 files in `/data/local/starbase/data/pds4/system_bundle/manifests` for you to verify. Moving forward, is this where you would like them?

**Comment by c-suh** [comment]
TODOs for self:
- update documentation regarding this script
- confirm location of files and where this script should be run

**Comment by rchenatjpl** [comment]
@c-suh , that's great. Thanks. With so many manifests, I suggest we break these into subdirectories, probably by node. Yes? The code should work anyway.

**Comment by c-suh** [comment]
@rchenatjpl

1. How about a subdirectory `collections` (or other, just lmk)? There are now 4 versions of the Collection files (8 files total), and this will only keep increasing.

2. Regarding subdirectories, there's currently a subdirectory for the year 2021. Would this be enough, or are you desiring more structure? If a yes to the latter...
   a. we'll have to run it by @jordanpadams as there'd be several considerations, and I'm sure I'm aware of only a few:
      - possibly upgrading the PDS4 Deep Archive to require a value from a fixed/restricted list and to write it into the manifest URL
      - letting the nodes know to update their version of the PDS4 Deep Archive (this one, we're already in the process of, so I don't think it'd be very different)
      - testing the new automated PDS4 SIP submitter for the NSSDCA for how it handles previously unaccounted for folders).

   b. how to merge a node subdirectory structure with calendar years - by years further organized by nodes, or by nodes first then by years?

**Comment by jordanpadams** [comment]
@c-suh @rchenatjpl unfortunately the directory structure, as-is, cannot change unless we symlink the files where you want them. the software has to hardcode this so NSSDCA knows where to go to find the SIPs/AIPs.

> How about a subdirectory collections (or other, just lmk)? There are now 4 versions of the Collection files (8 files total), and this will only keep increasing.

@c-suh if we want to create a separate "archive" or "past_versions" directory, that is fine. but the current collection must be in the location under the manifest directory.

> Regarding subdirectories, there's currently a subdirectory for the year 2021. Would this be enough, or are you desiring more structure? If a yes to the latter...

we unfortunately cannot do this. upgrading the software again will be a huge hassle. we can look into doing something like this down the road, but not in the foreseeable future. we need all SIPs/AIPs under a calendar year. per the node idea, the problem is there is no way to automate that. i'm trying to balance "clean but painfully manual" versus "simple and automated", but I almost always lean towards "simple and automated" wherever possible.

per the 4 versions of the collection files, maybe we can just write a quick wrapper script to run this once per week as a cron job, detect if there are new files, and then only re-generate the collection if something new exists?

we are getting a ton of these deliveries right now, but it will definitely settle over time.

**Comment by c-suh** [comment]
TLDR:
- no new subdirectories by node (sorry @rchenatjpl) 
- (to be settled) frequency and placement of the `makeCollection.py` script and the files it generates.

---

@jordanpadams, apologies; I need to better understand the components so that I can at least use the correct terminology. I intended the poorly named `collections` directory only for the files generated by Richard 's script (from https://github.jpl.nasa.gov/PDSEN/data-engineering/issues/131), e.g.
- Collection_product_aip_v#.0.csv
- Collection_product_aip_v#.0.xml
- Collection_product_sip_deep_archive_v#.0.csv
- Collection_product_sip_deep_archive_v#.0.xml

These hold information about the files en masse and, especially as these reports increase in quantity every week or however frequently the script is run, it seems tidier to place these separately. Again, I have spotty understanding of this, so let me know if this would disrupt anything or, if not, what to call this directory, e.g. `reports`, `info`, `labels`, `whoville`, `whatnot`?

Regarding frequency, I was running the script every time I added any bundles (what I mentally call the files from zips attached to issues with the 'nssdca-delivery' label), but can remove it from the [docs](https://pds-engineering.jpl.nasa.gov/content/nssdca_interface_process) and write a wrapper script to do it weekly only if new files are detected. @rchenatjpl any objections to this?

The bundles, I fully intend to leave as is in the `/data/pds4/manifests` or `/data/pds4/manifests/2021` directories, unless other subdirectories were to be created, which, if I'm understanding correctly, you have nixed. (I am also in favor of simple and automated.)

**Comment by rchenatjpl** [comment]
If I understand correctly and the following is ok, I vote to have only the latest version of each aip, sip, and Collection at the top level, then dump all the rest into one or more subdirectories (maybe one subdirectory per year). If we would need symlinks at the top level to the old ones, forget this.

**Comment by jordanpadams** [comment]
@c-suh copy. so the terms are a little funky here. the ZIP you are getting a actually just a set of AIPs/SIPs delivered to EN. bundles (or Product_Bundle in pds4 terminology) are what pds-deep-archive is ran against to produce those AIPs/SIPs.

per the comment regarding the generated collection .xml and .csv, the only catch is the latest version must be under `/data/pds4/manifests`. If it goes in any other directory, e.g. `/data/pds4/manifests/foobar`, that is no longer valid.

@rchenatjpl I am ok with putting old versions somewhere else, the only catch here (which I mention to many other nodes as well), is this now requires that we re-register the old versions so that we can point to the new path to that file. not the end of the world, but more work and software needed to make that happen. i am fine either way, just my two cents.

**Comment by rchenatjpl** [comment]
The actual locations of the AIPs and SIPs are registered somewhere? Yuk. This sounds like too much work just to declutter.

**Comment by jordanpadams** [comment]
@rchenatjpl 👍 they are not registered yet, but we will eventually do that.

**Comment by c-suh** [comment]
@jordanpadams aHAH, thank you for the explanations.

@rchenatjpl and Jordan, please confirm the following:

Decided: leave as is
- AIPs and SIPs (all versions)
- generated Collection .xml and .csv files (all versions)

TBD: frequency of generation Collection .xml and .csv files. Proposed suggestion is weekly, if there are new AIPs/SIPs. Current practice is whenever AIPs/SIPs are added.

**Comment by jordanpadams** [comment]
@c-suh i vote once per week.

**Comment by rchenatjpl** [comment]
I pass on voting on the frequency. I don't get what's in the 2021/ directory, but I think it would be fine to shunt off the non-current Collection* out of the main dir.

**Comment by c-suh** [comment]
@jordanpadams @rchenatjpl I've set up a cron job to do the following every Saturday at 11:59pm:
- compare the count (of AIP xml files) to the previous week's count
- if the number is different, run makeCollection.py and copy the latest version to the main directory

We agreed that previous versions of the generated files .xml and .csv files don't need to be in the main directory, but do they need to stay publicly accessible? There are already copies where the script is housed, so it seems redundant, unless you want to keep one as a backup directory, in which case, I'll create a sub-directory in https://pds.nasa.gov/data/pds4/manifests/ called `inventory` for these generated files

**Comment by jordanpadams** [comment]
@c-suh yes. they should remain online. as a I mentioned before, we are OK with not putting them in the directory, but the whole point with PDS4 data products is they must all be registered and available online at all time in case someone wants to go back and find a previous version. this applies to all PDS4 data, not just the AIPs/SIPs we are talking about here.

**Comment by jordanpadams** [comment]
otherwise sounds great!

**Comment by c-suh** [comment]
OK; the script will still run as specified above as above with 2 extra steps:
- copy the latest version of the generated files to https://pds.nasa.gov/data/pds4/manifests/inventory/
- remove the previous version's generated files from https://pds.nasa.gov/data/pds4/manifests/

Current directory structure of https://pds.nasa.gov/data/pds4/manifests/:
. **2021/**
... misc AIP/SIP sets
. **inventory/**
... _ALL_ generated collection .xml and .csv files
. misc AIP/SIP sets (from older version of Deep Archive tool)
. _LATEST_ generated collection .xml and .csv files
